x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/IndicesAccessControl.java

```diff
@@ -51,6 +54,22 @@ public boolean isGranted() {
         return granted;
     }
 
+    public void addPermissionsIfNotPresent(IndicesAccessControl other) {
+        final Map<String, IndexAccessControl> map = new HashMap<>(this.indexPermissions);
+        for (Map.Entry<String, IndexAccessControl> entry : other.indexPermissions.entrySet()) {
```

Comment:
Is it necessary to check `granted` before proceeding to merge?

Reply:
We should never get to this point with `granted` false, so yes, I think we need to throw an exception if `this.granted != other.granted`

Reply:
Thanks. After reading the code again, I realised that `IndicesAccessControl` with `false` grant is rejected by `AuthorizationResultListener` before it can reach `handleIndexActionAuthorizationResult`. So you are right, it should not happen. Thanks for the adding the check for extra safety.